### ISSUE TITLE
ParityDB corrupted on the RPC node

### ISSUE THREAD

**Issue by bakhtin** [paritytech/polkadot#7004]
I'm running a Polkadot RPC node (0.9.40-a2b62fb872b). The last upgrade was ~10 days ago (from v0.9.39 to v0.9.40) and the node was running just fine until yesterday. At around 7pm UTC April 3rd the following error appeared in the logs which now crashes the node each time it starts:
```
    2023-04-03 19:02:33 💤 Idle (62 peers), best: #14932334 (0x5093…5f24), finalized #14932331 (0x5269…582f), ⬇ 524.3kiB/s ⬆ 724.8kiB/s    
    2023-04-03 19:02:36 ✨ Imported #14932335 (0x9221…6fad)    
    2023-04-03 19:02:38 Accepting new connection 2/100
    2023-04-03 19:02:38 💤 Idle (62 peers), best: #14932335 (0x9221…6fad), finalized #14932332 (0xe517…cec6), ⬇ 456.6kiB/s ⬆ 580.0kiB/s    
    2023-04-03 19:02:42 ✨ Imported #14932336 (0xcd13…5235)    
    ====================
    Version: 0.9.40-a2b62fb872b
       0: sp_panic_handler::set::{{closure}}
       1: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
                 at rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/alloc/src/boxed.rs:2001:9
          std::panicking::rust_panic_with_hook
                 at rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panicking.rs:692:13
       2: std::panicking::begin_panic_handler::{{closure}}
                 at rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panicking.rs:579:13
       3: std::sys_common::backtrace::__rust_end_short_backtrace
                 at rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/sys_common/backtrace.rs:137:18
       4: rust_begin_unwind
                 at rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/panicking.rs:575:5
       5: core::panicking::panic_fmt
                 at rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/panicking.rs:65:14
       6: core::result::unwrap_failed
                 at rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/core/src/result.rs:1791:5
       7: <sp_state_machine::ext::Ext<H,B> as sp_externalities::Externalities>::storage
       8: sp_io::storage::get_version_1
       9: sp_io::storage::ExtStorageGetVersion1::call
      10: <F as wasmtime::func::IntoFunc<T,(wasmtime::func::Caller<T>,A1),R>>::into_func::wasm_to_host_shim
      11: <unknown>
      12: <unknown>
      13: <unknown>
      14: <unknown>
      15: <unknown>
      16: <unknown>
      17: <unknown>
      18: wasmtime_runtime::traphandlers::catch_traps::call_closure
      19: wasmtime_setjmp
      20: sc_executor_wasmtime::runtime::perform_call
      21: <sc_executor_wasmtime::runtime::WasmtimeInstance as sc_executor_common::wasm_runtime::WasmInstance>::call_with_allocation_stats

      22: sc_executor_common::wasm_runtime::WasmInstance::call_export
      23: sc_executor::native_executor::WasmExecutor<H>::with_instance::{{closure}}
      24: <sc_executor::native_executor::NativeElseWasmExecutor<D> as sp_core::traits::CodeExecutor>::call
      25: sp_state_machine::execution::StateMachine<B,H,Exec>::execute_aux
      26: sp_state_machine::execution::StateMachine<B,H,Exec>::execute_using_consensus_failure_handler
      27: <sc_service::client::client::Client<B,E,Block,RA> as sp_api::CallApiAt<Block>>::call_api_at
      28: <polkadot_runtime::RuntimeApiImpl<__SrApiBlock__,RuntimeApiImplCall> as sp_api::Core<__SrApiBlock__>>::__runtime_api_internal_call_api_at
      29: sc_transaction_pool::api::validate_transaction_blocking
      30: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
      31: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
      32: <futures_util::future::future::Map<Fut,F> as core::future::future::Future>::poll
      33: <tracing_futures::Instrumented<T> as core::future::future::Future>::poll
      34: tokio::runtime::task::raw::poll
      35: std::sys_common::backtrace::__rust_begin_short_backtrace
      36: core::ops::function::FnOnce::call_once{{vtable.shim}}
      37: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
                 at rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/alloc/src/boxed.rs:1987:9
          <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
                 at rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/alloc/src/boxed.rs:1987:9
          std::sys::unix::thread::Thread::new::thread_start
                 at rustc/90743e7298aca107ddaa0c202a4d3604e29bfeb6/library/std/src/sys/unix/thread.rs:108:17
      38: start_thread
      39: clone

Thread 'tokio-runtime-worker' panicked at 'Externalities not allowed to fail within runtime: "Trie lookup error: Database missing expected key: 0x390f741ca9dd8d47a7e07b562e9318024fe2b08069974bb1cdf87f8119675ca5"', /usr/local/cargo/git/checkouts/substrate-7e08433d4c370a21/ba87188/primitives/state-machine/src/ext.rs:192

This is a bug. Please report it at:
	https://github.com/paritytech/polkadot/issues/new
```

It seems that PartiyDB files have been corrupted in the runtime.

CLI arguments:
```
polkadot \
    --name=node-0 \
    --base-path=/data/ \
    --chain=polkadot \
    --database=paritydb \
    --pruning=archive \
    --prometheus-external \
    --prometheus-port 9615 \
    --rpc-external --ws-external --rpc-methods=safe --rpc-cors=all --telemetry-url='wss://submit.telemetry.parity-stg.parity.io/submit/ 1' \
    --public-addr=/ip4/${EXTERNAL_IP}/tcp/${RELAY_CHAIN_P2P_PORT} \
    --listen-addr=/ip4/0.0.0.0/tcp/${RELAY_CHAIN_P2P_PORT} \
    --listen-addr=/ip4/0.0.0.0/tcp/30333
```
You can download the corrupted DB archive from the GCS bucket (gs://parity-polkadot-backups/polkadot-paritydb-archive/poladot-public-sidecar-broken). The link expires in 5 days.

**Comment by bkchr** [comment]
CC @arkpar @cheme

**Comment by cheme** [comment]
When was the node first synch? I mean how old is the db?

**Comment by bakhtin** [comment]
More than 10 days ago. But I don't have exact dates. FWIW it was synced from the backup (gs://parity-polkadot-backups/polkadot-paritydb-archive)

**Comment by arkpar** [comment]
Is there a way to get a copy of the DB?

**Comment by cheme** [comment]
I list files with "gsutil ls -l gs://parity-polkadot-backups/polkadot-paritydb-archive/poladot-public-sidecar-broken/" , I think there is a cp command (currently don't have the disk size to copy), I don't remember how to login (I did in the past and it seems to be save).
I remember following instructions at https://github.com/paritytech/devops/wiki/Chain-Snapshots

**Comment by bakhtin** [comment]
The bucket is public, you don't need to login. You can download the contents with `gcloud storage cp -r gs://parity-polkadot-backups/polkadot-paritydb-archive/poladot-public-sidecar-broken <localdir>`

**Comment by arkpar** [comment]
Could it be that it ran out of disk space at some point?

**Comment by bakhtin** [comment]
The node has ~500GB of free disk space after it is fully synced. There were no changes to storage configuration in the past month.

**Comment by arkpar** [comment]
Do we have any other parity-db archive nodes configured in the same way? Are they doing fine so far?

**Comment by bakhtin** [comment]
Yes, I just spotted another node with the same issue. It is running in a Docker container though (paritypr/polkadot-debug:v0.9.40-rc8).

I just restored the first node from snapshot (gs://parity-polkadot-backups/polkadot-paritydb-archive/20230404-014542) and it is running fine so far.

**Comment by arkpar** [comment]
> Yes, I just spotted another node with the same issue. It is running in a Docker container though (paritypr/polkadot-debug:v0.9.40-rc8).

Did it happen on the same block as the first node?

**Comment by bakhtin** [comment]
The error references the same **key**, not sure if it is a reference to a block:
```
Thread 'tokio-runtime-worker' panicked at 'Externalities not allowed to fail within runtime: "Trie lookup error: Database missing expected key: 0x390f741ca9dd8d47a7e07b562e9318024fe2b08069974bb1cdf87f8119675ca5"', /usr/local/cargo/git/checkouts/substrate-7e08433d4c370a21/76fed9b/primitives/state-machine/src/ext.rs:192
```

**Comment by arkpar** [comment]
I've downloaded the database and was able to continue syncing with it using the latest polkadot master. This is probably a `find_entry_sse2` issue that was fixed in parity-db 0.4.6

**Comment by bkchr** [comment]
> Does it mean that upgrading to v0.9.41 should resolve the issue?

We only backported some commits and that didn't include the parity-db upgrade.
